allennlp/semparse/worlds/atis_world.py

```diff
@@ -0,0 +1,185 @@
+        return []
+
+    def all_possible_actions(self) -> List[str]:
+        """
+        Return a list of strings representing all possible actions
+        of the form lhs -> [rhs]
+        """
+        all_actions = set()
+        for non_term, rhs_list in self.valid_actions.items():
+            for rhs in rhs_list:
+                if non_term == 'string':
+                    all_actions.add('{} -> ["\'{}\'"]'.format(non_term, rhs))
+
+                elif non_term in ['number', 'asterisk', 'table_name']:
+                    all_actions.add('{} -> ["{}"]'.format(non_term, rhs))
+
+                else:
+                    ws_str = rhs.lstrip("(").rstrip(")")
+                    curr_child_strs = [tok for tok in re.split(" ws |ws | ws", ws_str) if tok]
+                    all_actions.add("{} -> [{}]".format(non_term, ", ".join(curr_child_strs)))
+
+        return sorted(all_actions)
+
+
+class SQLVisitor(NodeVisitor):
+    """
+    ``SQLVisitor`` performs a depths-first traversal of the the AST. It takes the parse tree
+    and gives us a action sequence that resulted in that parse.
+
+    """
+    def __init__(self, grammar: str) -> None:
+        """
+        Parameters
+        __________
+        grammar : ``str``
+            A string that descrbies the PEG (parsing expression grammar) in the form of:
+                nonterminal = ...
+        """
+        self.prod_acc: List[str] = []
+        self.grammar: Grammar = grammar
+
+        for nonterm in self.grammar.keys():
+            if nonterm != 'stmt':
+                self.__setattr__('visit_' + nonterm, self.add_prod_rule)
+
+    def generic_visit(self, node: Node, visited_children) -> None:
+        self.add_prod_rule(node)
+
+    def add_prod_rule(self, node: Node, children=None) -> None: # pylint: disable=unused-argument
+        """
+        For each node, we accumulate the rules that generated its children in a list
+        """
+        if node.expr.name and node.expr.name != 'ws':
+            lhs = '{} -> '.format(node.expr.name)
+
+            if isinstance(node.expr, Literal):
+                rhs = '["{}"]'.format(node.text)
+
+            else:
+                child_strs = []
+                for child in node.__iter__():
+                    if child.expr.name == 'ws':
+                        continue
+                    if child.expr.name != '':
+                        child_strs.append(child.expr.name)
+                    else:
+                        ws_str = child.expr._as_rhs().lstrip("(").rstrip(")") # pylint: disable=protected-access
+                        curr_child_strs = [tok for tok in re.split(" ws |ws | ws", ws_str) if tok]
+                        child_strs.extend(curr_child_strs)
+                rhs = "[" + ", ".join(child_strs) + "]"
+
+            rule = lhs + rhs
+            self.prod_acc = [rule] + self.prod_acc
+
+    def visit_stmt(self, node: Node, children=None) -> List[str]: # pylint: disable=unused-argument
```

Comment:
So for each production rule it will either call `visit_{nonterminal}` if it is defined, or `generic_visit`.  The starting symbol in my grammar was `stmt` so it was calling `visit_stmt`. I'll avoid this by just calling `generic_visit` for all my nodes, instead of `visit_{nonterminal}` and using the `@overrides` decorator there.